accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,431 @@
+
+```diff
+namespace System.Reflection
+{
++   public static class MetadataLoadContextExtensions
++   {
+        // These support function pointer types on FieldInfo.FieldType, PropertyInfo.PropertyType and ParameterInfo.ParameterType.
+        // InvalidOperationException is thrown when the corresponding type.IsFunctionPointer==false.
++       public static Type GetSignatureType(this FieldInfo fieldInfo);
++       public static Type GetSignatureType(this PropertyInfo propertyInfo);
++       public static Type GetSignatureType(this ParameterInfo parameterInfo); 
+        // or "GetFunctionPointerSignatureType()?
+
+        // These only work for types obtained from GetSignatureType().
+        // An empty Type[] is returned for non-signature function pointer types.
+        // InvalidOperationException is thrown when type.IsFunctionPointer==false.
++       public static Type[] GetRequiredCustomModifiers(this Type);
+        // or "GetSignatureTypeRequiredCustomModifiers()?
++       public static Type[] GetOptionalCustomModifiers(this Type);
+        // or "GetSignatureTypeOptionalCustomModifiers()?
++   }
+}
+```
+
+Sample usage:
+```c#
+public class Holder
+{
+    public unsafe delegate*<in int, out int, void> _field;
+}
+
+FieldInfo fieldInfo = typeof(Holder).GetField("_field");
+
+// Get the signature type from the field; this Type is like fieldInfo.FieldType but it includes custom mods
+Type fnType = fieldInfo.GetSignatureType();
+
+// Get the function pointer's first parameter's type
+Type[] p1Type = fnType.GetFunctionPointerParameterTypes()[0];
+
+// Get the required modifiers from the signature type
+Type[] p1ReqMods = p1Type.GetRequiredCustomModifiers();
+
+// The required modifiers should include the 'in' modifier
+Debug.Assert(p1ReqMods[0].GetType() == typeof(Runtime.InteropServices.InAttribute));
+```
+
+## System.Reflection.Emit.ILGenerator
+See the existing [ILGenerator.EmitCalli()](https://docs.microsoft.com/dotnet/api/system.reflection.emit.ilgenerator.emitcalli#system-reflection-emit-ilgenerator-emitcalli) for reference.
+```diff
+namespace System.Reflection.Emit
+{
+    public class ILGenerator
+    {
++       public virtual void EmitCalli(
+            // Must be OpCodes.Calli which matches existing EmitCalli convention
+            // Not sure why this is redundant (for future Calli opcodes?)
++           System.Reflection.Emit.OpCodes opcode,
++           Type[] callingConventions,
++           Type? returnType,
++           Type[]? parameterTypes);
+
++       public virtual void EmitCalli(
++           System.Reflection.Emit.OpCodes opcode,
++           Type functionPointerType)
+    }
+}
+```
```

Comment:
Reflection.Emit will need some work to deal with function pointers correctly beyond the new EmitCalli overloads.

Consider this example that tries to emit a call to a method that takes function pointer https://gist.github.com/jkotas/79c7c673b7d892f41f6be5f16d380aed . It does work today. MethodInfo returns IntPtr instead of actual type, Reflection.Emit encodes method reference with a wrong signature, and we get MissingMethodException at runtime.

The problem is that this example cannot work even once this proposal is implemented. The example contains a function pointer with modreqs, but Reflection.Emit has no way to see the modreqs. It means that it is still going to emit reference with a wrong signature and we will get MissingMethodException at runtime.

This suggests that we cannot actually omit function pointer modopts/modreqs in runtime reflection object model if we want to have full support for function pointers in Reflection.Emit.

Reply:
Wouldn't we run into the same problem if we wrap access to a `get; init;` setter like that? I think the problem is really that the runtime reflection system was not designed to introspect modifiers. It was designed to ignore them.